Added target_group_arns to ASG configuration

aivanov1 opened the pull request.

aivanov1 commented:
Tried using the **target_group_arns** parameter of the ASG, but it didn't work for me. Hence, adding a single Target Group as an attachment. In case there is no Target Group specified to the module, it still works since it's an optional parameter.
This is regarding issue:
https://github.com/hashicorp/terraform-aws-consul/issues/69

aivanov1 commented on modules/consul-cluster/main.tf:
```diff
@@ -39,10 +39,15 @@ resource "aws_autoscaling_group" "autoscaling_group" {
 
+resource "aws_autoscaling_attachment" "consul-asg-tg-attachment" {
+  autoscaling_group_name = "${aws_autoscaling_group.autoscaling_group.name}"
+  alb_target_group_arn   = "${var.target_group_arn}"
```
The "alb_target_group_arn" argument is [optional](https://www.terraform.io/docs/providers/aws/r/autoscaling_attachment.html). I tried it and the terraform created an attachment without the argument (see output below). Note the attachment is still referenced by the ASG, it just doesn't contain any value. Also verified the field *Target Groups* was empty in the AWS Console ASG UI. Note this field is a list, so since we don't add anything to it, I expect it will remain an empty list.
`  + module.prod_consul_1.module.consul_servers.aws_autoscaling_attachment.consul-asg-tg-attachment
      id:                                        <computed>
      autoscaling_group_name:                    "${aws_autoscaling_group.autoscaling_group.name}"
---> NOTE "alb_target_group_arn" is missing from here, which is where it shows up if specified

  + module.prod_consul_1.module.consul_servers.aws_autoscaling_group.autoscaling_group
      id:                                        <computed>
      arn:                                       <computed>
      default_cooldown:                          <computed>
      desired_capacity:                          "3"
      ...
      target_group_arns.#:                       <computed>
      ...
`

aivanov1 commented:
I realized an issue with this solution is one may want to configure multiple Target Group's for ALB/NLB (NLB also uses target groups), e.g. one for HTTP (port 8500) and one for DNS (port 8600). In this case the 1 attachment of a single Target Group would not work.
Hence, I went back to my initial attempt to get **target_group_arns** argument on the ASG to work, and I figured out what I was doing wrong - I didn't pass the parameter as **["${param}]** (note the square brackets). Submitting a commit with the change shortly...

brikis98 commented:
Hm, we actually removed `target_group_arns` in https://github.com/hashicorp/terraform-aws-consul/pull/57. See discussion in https://github.com/hashicorp/terraform-aws-nomad/issues/19 as to why.

aivanov1 commented:
Well, the current code doesn't seem to work since once the infrastructure with an _autoscaling_group_attachment_ is created if we run *terraform apply* again, we runs into:
**Terraform v0.11.7**
```
Terraform will perform the following actions:

  ~ module.prod_consul_1.module.consul_servers.aws_autoscaling_group.autoscaling_group
      target_group_arns.#:         "1" => "0"
      target_group_arns.842221997: "arn:aws:elasticloadbalancing:us-east-2:423487953385:targetgroup/my-lb-tg/2f4d7259f471d85a" => ""
```

brikis98 commented:
Not sure I follow. Are you saying that: 

1. You add an `aws_autoscaling_attachment` resource to your code.
1. Run `apply`.
1. The attachment is created and you're able to use the ALB to talk to Consul.
1. You run `apply` again and Terraform removes the attachment?

aivanov1 commented:
Here's what happened:
1. The code I was testing was using v0.2.2, which contains `target_group_arns` argument:
https://github.com/hashicorp/terraform-aws-consul/blob/0e5a7b15cc3b2cdf03201381696842c6f8310c38/modules/consul-cluster/main.tf
2. I added an aws_autoscaling_attachment in my code, which apparently created a conflict with this argument. Looking at the [aws_autoscaling_attachment](https://www.terraform.io/docs/providers/aws/r/autoscaling_attachment.html) doc, they've mentioned the conflict with `load_balancers` argument, but probably forgot to note the `target_group_arns` argument.
3. When I looked at the `terraform-aws-consul` code, I should have looked at v0.2.2. Instead I looked at the latest code, and since the argument is already removed, I got the impression ASG attachments clash with the module, which is **not** true for v0.3.5 (latest).

I switched to v0.3.5 (latest), and the issue went away. Thank you for your help figuring this out!
